pkg/cloud/openstack/machineset/cache.go

```diff
@@ -0,0 +1,82 @@
+package machineset
+
+import (
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/flavors"
+	"sync"
+	"time"
+)
+
+const StaledTime time.Duration = 300 * time.Second
+const RefreshFailureTime time.Duration = 60 * time.Second // This controls how often we try to get a look at a failed flavor
+// machineFalvorKey is used to identify Machine flavor
+type machineFlavorKey struct {
+	name string
+}
+
+type flavorCacheEntry struct {
+	flavorInfoPtr *flavors.Flavor
+	updateTime    time.Time
+	flavorName    string
+}
+
+//machineFlavorCache is used for caching machine flavors
+type machineFlavorsCache struct {
+	cacheMutex sync.Mutex
+	cache      map[machineFlavorKey]flavorCacheEntry
+}
+
+func newMachineFlavorCache() *machineFlavorsCache {
+	return &machineFlavorsCache{
+		cache: map[machineFlavorKey]flavorCacheEntry{},
+	}
+}
+
+func (mfc *machineFlavorsCache) getFlavorInfo(osService OpenStackInstanceService, flavorName string) *flavors.Flavor {
+	// We don't return errors. If we can't look up the flavor we simply return nil.
+
+	mfc.cacheMutex.Lock()
+	defer mfc.cacheMutex.Unlock()
+	if entry, ok := mfc.cache[machineFlavorKey{name: flavorName}]; ok {
+		// An entry in the cache has been found but we still need to figure out if it is valid
+
+		if entry.flavorInfoPtr != nil && time.Now().Sub(entry.updateTime) < StaledTime {
+			// We have current valid entry
+			return entry.flavorInfoPtr
+		}
+		if entry.flavorInfoPtr == nil && time.Now().Sub(entry.updateTime) < RefreshFailureTime {
+			// We have an invalid entry but is too soon to try an refresh, so we return nil
+			return entry.flavorInfoPtr
```

Comment:
```suggestion
			return nil
```
I know the pointer is nil here, but for the sake of clarity, if you always return nil here, then the return statement should say that

Reply:
done